rpcs3/Emu/RSX/rsx_cache.h

```diff
@@ -505,39 +513,97 @@ namespace rsx
 			}
 
 			dlg->create();
+			dlg->set_limit(0, entry_count);
+			dlg->set_limit(1, entry_count);
+			dlg->update_msg(0, 0, entry_count);
+			dlg->update_msg(1, 0, entry_count);
+
+			// Setup worker threads
+			unsigned nb_threads = std::thread::hardware_concurrency();
+			std::vector<std::thread> worker_threads(nb_threads);
 
-			const auto prefix_length = version_prefix.length();
-			u32 processed = 0;
-			while (root.read(tmp) && !Emu.IsStopped())
+			// Preload everything needed to compile the shaders
+			// Can probably be parallelized too, but since it's mostly reading files it's probably not worth it
+			std::vector<std::tuple<pipeline_storage_type, RSXVertexProgram, RSXFragmentProgram>> unpackeds;
+			std::chrono::milliseconds last_update;
+			u32 processed_since_last_update = 0;
+			for (u32 i = 0; (i < entry_count) && !Emu.IsStopped(); i++)
 			{
-				if (tmp.name == "." || tmp.name == "..")
-					continue;
+				fs::dir_entry tmp = entries[i];
 
 				const auto filename = directory_path + "/" + tmp.name;
 				std::vector<u8> bytes;
 				fs::file f(filename);
-
-				processed++;
-				dlg->update_msg(processed, entry_count);
-
 				if (f.size() != sizeof(pipeline_data))
 				{
 					LOG_ERROR(RSX, "Cached pipeline object %s is not binary compatible with the current shader cache", tmp.name.c_str());
 					invalid_entries.push_back(filename);
 					continue;
 				}
-
 				f.read<u8>(bytes, f.size());
+
 				auto unpacked = unpack(*(pipeline_data*)bytes.data());
-				m_storage.add_pipeline_entry(std::get<1>(unpacked), std::get<2>(unpacked), std::get<0>(unpacked), std::forward<Args>(args)...);
+				m_storage.preload_programs(std::get<1>(unpacked), std::get<2>(unpacked));
+				unpackeds.push_back(unpacked);
 
-				tally += delta;
-				if (tally > 1.f)
+				// Only update the screen at about 10fps since updating it everytime slows down the process
+				std::chrono::milliseconds now = std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::system_clock::now().time_since_epoch());
```

Comment:
Pretty sure you dont need time since epoch, just a forward incrementing counter. Consider just using steady_clock::now()